### PR TITLE
Replace backend with fsspec

### PR DESCRIPTION
**Pull request by baxen**
This is a major refactor but should come with improved performance and a clear path forward to support other cloud filesystems. Starting with a beta release to get some testing in real projects

See changelog for more details.

### CODE REVIEW

File: blocks/filesystem/base.py [hunk 1]
```diff
@@ -21,62 +48,128 @@ def ls(self, path: str) -> Sequence[str]:
...
+            fs = self._get_filesystem(pd)
+            fs.put(sources, dest, recursive=recursive)
+
+        if pd is None:
+            fs = self._get_filesystem(ps)
+            fs.get(sources, dest, recursive=recursive)
+
+        if pd == "gs":
+            if isinstance(dest, str):
+                fs.invalidate_cache(dest)
```
Comment (stillmatic) [inline blocks/filesystem/base.py:112]: why is this necessary?

File: blocks/core.py [hunk 2]
```diff
@@ -20,22 +20,24 @@
...
 def assemble(
     path: str,
     cgroups: Optional[Sequence[cgroup]] = None,
     rgroups: Optional[Sequence[rgroup]] = None,
     read_args: Any = {},
     cgroup_args: Dict[cgroup, Any] = {},
     merge: str = "inner",
-    filesystem: FileSystem = GCSFileSystem(),
+    filesystem: FileSystem = FileSystem(),
+    tmpdir: str = None,
```
Comment (stillmatic) [inline blocks/core.py:40]: nit: Optional[str]

File: blocks/filesystem/base.py [hunk 3]
```diff
@@ -1,18 +1,45 @@
-from abc import ABC, abstractmethod
+from typing import Union, Sequence, Tuple, List
+from collections import defaultdict
+from fsspec.core import split_protocol, get_filesystem_class, has_magic
 
-from typing import Union, IO, Sequence, Iterator, List, Any, Generator, ContextManager
-from blocks.datafile import DataFile
-from contextlib import contextmanager
 
+class FileSystem:
```
Comment (stillmatic) [inline blocks/filesystem/base.py:6]: super nit: this class name is now a bit of a misnomer since it can contain multiple fsspec filesystems

File: blocks/filesystem/base.py [hunk 4]
```diff
@@ -1,18 +1,45 @@
...
+        Additional options passed to each filesystem for each protocol
+        e.g. {'gs': {'project': 'example'}} to set the gs filesytem project to example
     """
 
-    @abstractmethod
+    def __init__(self, **storage_options):
+        self.storage_options = defaultdict(dict)
+        self.storage_options.update(storage_options)
+        self.storage_options[None]["auto_mkdir"] = True
+        self.filesystems = {}
```
Comment (stillmatic) [inline blocks/filesystem/base.py:24]: nit: would be nice to instantiate the filesystem at this time with the given storage options, if present

File: blocks/dfio.py [hunk 5]
```diff
@@ -33,34 +29,23 @@ def read_df(datafile, **read_args):
...
+    return reader(path, **read_args)
 
 
-def write_df(df, datafile, **write_args):
+def write_df(df, path, **write_args):
     """Write a dataframe to file based on the file extension
 
     The following formats are supported:
-    parquet, avro, csv, pickle
+    parquet, avro, csv, pickle, json
```
Comment (stillmatic) [inline blocks/dfio.py:48]: would be nice to have a note on which types are supported. reading the code, should be any writer supported by pandas (i.e. `pandas.DataFrame.to_xyz`), with some possibly additional glue code.

File: blocks/core.py [hunk 6]
```diff
@@ -239,29 +232,32 @@ def partitioned(
...
+            frames.append(read_df(p, **args))
         return _merge_all(frames, merge=merge)
 
     for rgroup in _shared_rgroups(grouped):
         blocks.append(merged(rgroup))
     return dd.from_delayed(blocks)
 
 
+@with_function_tmpdir
 def place(
```
Comment (stillmatic) [inline blocks/core.py:255]: seems pretty similar to https://filesystem-spec.readthedocs.io/en/latest/features.html#remote-write-caching ?

**Review by stillmatic — CHANGES_REQUESTED** [review]
looks great, makes a lot of sense and should be easier to use, esp with s3. a couple of exciting features in fsspec are pyarrow integration and transactions, thanks for working on this!

main issue is adding `wrapt` as an install dependency in `setup.py` (CI is failing). otherwise, some very minor nits inline.

File: blocks/filesystem/base.py (hunk 1 above)
Comment (baxen) [reply blocks/filesystem/base.py:112]: gcsfs (and probably some of the other implementations) cache many of their operations so future calls to list that destination will potentially not reflect the newly added files. So i am trying to cache invalidate whenever we add new data to the remote filesystem

I'll add an explicit comment here to make this clear to future contributors

File: blocks/core.py (hunk 2 above)
Comment (baxen) [reply blocks/core.py:40]: yeah i'm confused what to do here with typehints/docstring, this is not actually optional at all but is provided by the decorator instead so the user should never pass it in

File: blocks/core.py (hunk 6 above)
Comment (baxen) [reply blocks/core.py:255]: Yeah there's definitely a lot of overlap, but this does handle a copy of other details including using a local path (not a file handle) which is more compatible with pandas functions and infering the pandas function to use based on file extension

**Review by stillmatic — APPROVED** [review]

File: blocks/dfio.py [hunk 7]
```diff
@@ -12,16 +10,14 @@
     avro_imported = False
 
 
-def read_df(datafile, **read_args):
-    """Read a dataframe from file based on the file extension
-
-    The following formats are supported:
-    parquet, avro, csv, pickle
+def read_df(path, **read_args):
```
Comment (lexi-squid-ding) [inline blocks/dfio.py:13]: nice, this is a lot more readable

File: blocks/dfio.py [hunk 8]
```diff
@@ -75,71 +60,49 @@ def write_df(df, datafile, **write_args):
...
+    if write_name == "to_csv":
+        # make index=False the default for similar behaviour to other formats
+        write_args["index"] = write_args.get("index", False)
 
-def _read_avro(handle, **read_args):
+    write_fn = getattr(df, write_name)
+    write_fn(path, **write_args)
+
+
+def _read_avro(path, **read_args):
```
Comment (lexi-squid-ding) [inline blocks/dfio.py:91]: nit: are you passing the **read args?

File: blocks/filesystem/gcs_filesystem.py [hunk 9]
```diff
@@ -1,236 +1,14 @@
...
-    """Create a tempdir that will be cleaned up at session exit"""
-    tmpdir = tempfile.mkdtemp()
-    # create and use a subdir of specified name to preserve cgroup logic
-    atexit.register(lambda: shutil.rmtree(tmpdir))
-    return tmpdir
+    """Deprecated, see FileSystem"""
+
+    def __init__(self, *args, **kwargs):
+        warnings.warn(
+            "This class is deprecating, use blocks.filesystem.FileSystem",
```
Comment (lexi-squid-ding) [inline blocks/filesystem/gcs_filesystem.py:11]: nit: "deprecated"

**Review by lexi-squid-ding — APPROVED** [review]
LGTM aside from some nits!

File: blocks/utils.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import atexit
+import tempfile
+
+import wrapt
+
+
+@wrapt.decorator
+def with_function_tmpdir(wrapped, instance, args, kwargs):
```
[the remark this answers is not in the document]
Comment (baxen) [reply blocks/utils.py:8]: ah this is provided by `wrapt` in all decorators but we do not need or use it. just following their naming structure for consistency

File: blocks/dfio.py (hunk 8 above)
Comment (baxen) [reply blocks/dfio.py:91]: in this case no additional read args are implemented, but left them here for consistency and if we add any in the future

**Comment by baxen** [comment]
We've been beta testing for a few months now and this seems like a conclusive improvement. Will release this as 0.9.0
